formula/src/test/java/com/instacart/formula/subjects/MultiThreadRobot.kt

```diff
@@ -0,0 +1,87 @@
+package com.instacart.formula.subjects
+
+import com.google.common.truth.Truth
+import com.instacart.formula.subjects.SleepFormula.SleepEvent
+import com.instacart.formula.test.TestableRuntime
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.TimeUnit
+
+class MultiThreadRobot(val runtime: TestableRuntime) {
+    class NamedThreadFactory(private val name: String): ThreadFactory {
+        override fun newThread(r: Runnable): Thread {
+            return Thread(r, name)
+        }
+    }
+
+    private val threadFormula = SleepFormula()
+    private val observer = runtime.test(threadFormula, "initial-key")
+
+    // Manage executors and update completion
+    private val executorMap = mutableMapOf<String, Executor>()
+    private val eventCompletionLatches = ConcurrentLinkedQueue<CountDownLatch>()
+
+    fun thread(name: String, sleepDuration: Long) = apply {
+        thread(name) {
+           observer.output {
+               this.onSleep(sleepDuration)
+           }
+        }
+    }
+
+    fun thread(name: String, function: () -> Unit) = apply {
+        val executor = executorMap.getOrPut(name) {
+            Executors.newSingleThreadExecutor(NamedThreadFactory(name))
+        }
+
+        // Creating a latch and adding it to a list to make sure we are able to
+        // wait for all event completion.
+        val completionLatch = CountDownLatch(1)
+        eventCompletionLatches.add(completionLatch)
+
+        executor.execute {
+            observer.output {
+                function()
+                completionLatch.countDown()
+            }
+        }
+    }
+
+    fun input(newKey: String) = apply {
```

Comment:
This comment is not related to this line. So we're handing off Formula update events to the currently processing Thread, which might be a network/IO thread. I'm not convinced this is the most effective use of our resources. Background network thread is designed to take care of network calls boilerplate, it's not conceptually designed to process app UI state management updates. I think the further network threads usually go is up to parsing response to a more digestible data. 

I'm curious if we should explore Formula runtime having 1 working background thread that runs transitions and side effects. If a side effect needs to run on a main thread, there would be an explicit API to do so, such as:
```
transition(state = state.copy(...)) {
    uiThread {
        ...    
    }
    backgroundThread {
        ...    
    }
}
```

Reply:
Good point about IO-thread. So, we basically have 2 options: 
- don't perform any enforcement and let user switch thread manually via in the `RxAction.fromObservable{}` block via `observeOn(computation())`.
- or maybe let formula enforce thread. E.g. we can introduce background thread (or a pool of threads) and if the event received not on UI thread and not on one of our background threads, we execute it on the background thread from the pool.

I would prefer the second option.

The API to request on which thread perform side effect is useful. I only see one downside: it could add mental overhead to the formula user and code-reviewer. Engineers will need to spend additional time to decide/verify what should be on `uiThread` what should be on `backgroundThread`. I can be useful though.

Reply:
I might be wrong, but I’m not sure the statement that “network/IO thread is not conceptually designed to process app UI state management updates” is correct. The thread classification such as I/O or computation only matters in terms of thread pools (and how many threads are allocated). I’m not sure there is a problem that a network thread executes more code (and will exit shortly after). Not having to jump a thread to process the update has two benefits:
- We process the formula update much faster (jumping to the main thread can have a huge 100+ms delay; I’ve seen 300+ms during app start)
- We potentially defer some processing to a background thread, leaving the main thread more responsive

> I'm curious if we should explore Formula runtime having 1 working background thread

The problem with using a single background worker is that each user event will have to jump from main thread to background thread and then back to the main thread. The delay added by this can make things feel less responsive. This is why I think better option is not to confine the formula to a single thread which would require constant jumping back and forth. The current approach tries to execute things on the thread that has the event.

> If a side effect needs to run on a main thread, there would be an explicit API to do so, such as:

In terms of extra threading APIs for side-effect execution, it might be convenient, but I'm not sure it's the right approach. Currently, the formula doesn't manage any threads and doesn't know about the main thread. Not sure introducing those for this use-case is worth it. Regardless if we introduce this threading API within side-effects, the end user will still have to think much more carefully about shared mutable state and how to synchronize reads/writes. I suspect logic to ensure thread-safety should live within classes that we call the side-effects on instead of formula.